### DISCUSSION TITLE
ADS-B and Mode S decoding ✈️

### DISCUSSION THREAD

**Discussion by xoolive** [xoolive/traffic#130] [Ideas]
Let's use this thread to discuss about decoding capabilities for the raw data obtained from the Impala shell or directly from your own ADS-B receiver. 

[pyModeS](https://github.com/junzis/pyModeS) may be your library of choice: that's why it is directly integrated in traffic.

**Comment by ElSabio97** [comment]
Hi @xoolive and @junzis !

I wanted to ask you for some extra-advice.

What would be your approach to getting RA (resolution advisories) events and then plotting some of them?

At this moment I do the following: 

1. I edited the `raw` query type from @junzis `pyopensky` `OpenskyImpalaWrapper` to download from the `acas_data4` table instead of the `rollcall_replies_data4`.
2. I download aproximatelly 200.000 `acas_data4` messages.
3. Then the resulting data frame's raw messages column is processed to get which ones are BDS 30. This is done to get the threat's `icao24.`
4. Now, with the few raw messages that are BDS30 and also give the threat's icao24, I manually choose an icao24 and the threat's icao24. 
5. I write both icao24 addreses in the `icao24` filter argument of a `history` query using a list. Also start/stop datetime filter arguments are set 10 minutes before and after the RA datetime.
5. I export the query dataframe to a pickle file.
6. With that done, I go to the traffic `GUI`, select the pickle file and then plot the trajectories, altitudes and vertical rates of the two planes.

It's a long process I know, but this is the best I have done so far.

The thing is I recently learned while reading through the ICAO DOC 9871 that there are other ways to get the RA information. As far as I'm concerned: BDS10, BDS30, BDS61, BDS65 and on the ADSB's TypeCode 28 SubType 2.

And now I guess the best way to get this information is by using the `rawdata` or `extended` query not by using the `acas_data4` table because it won't be there since I believe it only contains `DF16` messages. And then using `pyModeS` functions to decode the messages.

But as you know, raw messages downloads take forever, even if there is a bound for an airport or shape. I have been searching and reading about how to minimize the download time with no luck.

Sorry for the longest comment on github history.

**Reply by junzis** [reply]
Unfortunately, I think there is no faster solution. Since Mode S information is not decoded by OpenSky, you have to download the raw messages and decoded them locally. This process always takes a very long time. You have to be selective in terms of the time scale for your study.

**Reply by ElSabio97** [reply]
Thank you for your reply!

I was wondering how @xoolive  & Co. got it [on this article](http://www.cs.ox.ac.uk/files/11077/DASC_2019.pdf).

I suppose the download process took days. And the funny thing is that I could get my hands on the .csv used on that article (If you read this, thanks again Martin), but it only weighs 45MB!

**Reply by xoolive** [reply]
For the DASC paper, the decoding was done by Matthias with the java parser for the paper, directly from the raw data feed.

**Reply by stinerk** [reply]
I also want to get RA data and do the work you did. I don't know what to do because I don't have knowledge like decoding. Could you provide me with your modified raw, pyopensky, and openskyimplawrapper? Also, I would appreciate it if you could tell me how I can demonstrate and view the materials I received like that. I really want data, but I don't know how to look at them. I'm desperate for your help
